sdks/java/core/src/main/java/org/apache/beam/sdk/util/PubsubClient.java

```diff
@@ -16,22 +16,90 @@
+
+import com.google.api.client.util.Clock;
+import com.google.api.client.util.DateTime;
+import com.google.common.base.Preconditions;
 
 import java.io.IOException;
 import java.io.Serializable;
-import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
 
 /**
- * A helper interface for talking to Pubsub via an underlying transport.
+ * An (abstract) helper class for talking to Pubsub via an underlying transport.
  */
-public interface PubsubClient extends AutoCloseable {
+public abstract class PubsubClient implements AutoCloseable {
+  /**
+   * Which underlying transport to use.
+   */
+  public enum TransportType {
+    APIARY,
+    GRPC;
+  }
+
+  /**
+   * Return timestamp as ms-since-unix-epoch corresponding to {@code timestamp}.
+   * Return {@literal null} if no timestamp could be found. Throw {@link IllegalArgumentException}
+   * if timestamp cannot be recognized.
+   */
+  @Nullable
+  private static Long asMsSinceEpoch(@Nullable String timestamp) {
+    if (timestamp == null || timestamp.isEmpty()) {
+      return null;
+    }
+    try {
+      // Try parsing as milliseconds since epoch. Note there is no way to parse a
+      // string in RFC 3339 format here.
+      // Expected IllegalArgumentException if parsing fails; we use that to fall back
+      // to RFC 3339.
+      return Long.parseLong(timestamp);
+    } catch (IllegalArgumentException e1) {
+      // Try parsing as RFC3339 string. DateTime.parseRfc3339 will throw an
+      // IllegalArgumentException if parsing fails, and the caller should handle.
+      return DateTime.parseRfc3339(timestamp).getValue();
+    }
+  }
+
+  /**
+   * Return the timestamp to use for a Pubsub message with {@code attributes}:
+   * <ol>
+   * <li> If {@code timestampLabel} is non-{@literal null} then the message attributes
+   * must contain that label, and the value for that label must be recognized as a
+   * ms-since-unix-epoch or RFC3339 time. Return that time as ms-since-unix-epoch or
+   * throw {@link IllegalArgumentException} if not possible.
+   * <li> Otherwise return the current system time according to {@code clock}.
+   * </ol>
+   * <p>
+   * Note  that {@code pubsubTimestamp} is currently ignored, but should be the Pubsub timestamp
+   * extracted from the underlying Pubsub message. This may be used for the message timestamp
+   * instead of the current system time in the future.
+   *
+   * @throws IllegalArgumentException
+   */
+  protected static long extractTimestamp(
+      Clock clock,
+      @Nullable String timestampLabel,
+      @Nullable String pubsubTimestamp,
+      @Nullable Map<String, String> attributes) {
+    if (timestampLabel == null) {
+      return clock.currentTimeMillis();
+    }
+    Long timestampMsSinceEpoch =
+        asMsSinceEpoch(attributes == null ? null : attributes.get(timestampLabel));
+    Preconditions.checkArgument(timestampMsSinceEpoch != null,
```

Comment:
static import `checkArgument` and other functions from `Preconditions`, and make this code clearer.